### PR TITLE
Fix slice and get info.

### LINKED ISSUE #5018: xgb.cv doesn't return features names
Hi all, 

Long fan of your efforts with the Xgboost algorithm/implementation. It is super fast and memory-friendly.

I found a problem when trying to see feature importance when using the `xgb.cv` function, namely that it doesn't return the features names when using the callback `cb.cv.predict(save_models = TRUE)`. 

I found this trying to plot the model importance using  `xgb.plot.importance`.
Does the numbers refer to the python way of counting columns (i.e., starting from 0)?

I made an MRE below: 

Xgboost version: xgboost_0.90.0.2 (R package)

```R
data(iris)

iris <- filter(iris, Species != 'setosa')
features <- as.matrix(iris[, !grepl('Species', colnames(iris))])
label <- ifelse(iris$Species == 'virginica', 1, 0)

library(xgboost)

model <- xgboost::xgb.cv(
                      data = features
                    , label = label 
                    , nfold = 5 
                    , nrounds = 25
                    , metrics = list("auc")
                    , stratified = TRUE
                    , verbose = TRUE
                    , callbacks = list(cb.cv.predict(save_models = TRUE))
                    
                    , params = list(
                        eta = 0.1
                        , max_depth = 10
                        , objective = "binary:logistic"
                        , colsample_bytree = 0.5
                        , subsample = 0.5
                        , nthread = 2
                        , seed = 1
                        )
                    )

importance <- xgb.importance(model = model$models[[1]])
xgboost::xgb.plot.importance(importance)

```
xgboost_reprex

### LINKED ISSUE #5288: [R] base_margin not working in a Multiclass Cross-validation
Firstly, I see this topic has been previously raised/closed here [#2006](https://github.com/dmlc/xgboost/issues/2006), but I am still having the issue when running a multiclass classifier.  For reference I am using xgBoost version 1.0.0.1 via R:

I cannot get the base_margin passed through during cross-validation for a multiclass model.  I can get it working on xgb.train using the exact same code (save for the **nfold** argument), but the base_margin is not pass to the cross-validation run.

Further, the base_margin does pass through to the Cross Validation when when I run a standard binary classifier using the agaricus data (I can provide my code for this if needed).  It is just the multiclass scenario that doesn't work.

I have provided some code that should hopefully be replicated for the multi-class case.  

    #load xgboost
    library(xgboost)
     
    #define number of classes, features and label
    num_class <- 3
    data = as.matrix(iris[, -5])
    label =  as.numeric(iris$Species) - 1

    #create xgbMatrix
    xgbMat <- xgb.DMatrix(data = data, label = label)
    watchlist = list(train = xgbMat)

    #run initial boost
    bst <- xgb.train(data = xgbMat,
                     max_depth = 4, eta = 0.4, nrounds = 50,
                     objective = "multi:softprob", eval_metric = "mlogloss",
                     watchlist = watchlist,
                     num_class = num_class, print_every_n = 10)

That achieves the following multiclass log-loss metrics:

    [1]  train-mlogloss:0.636670 
    [11] train-mlogloss:0.039032 
    [21] train-mlogloss:0.022351 
    [31] train-mlogloss:0.018615 
    [41] train-mlogloss:0.016476 
    [50] train-mlogloss:0.015429

So the best log-loss is 0.015429

Now I predict from this boost and use the margin output as the base margin in a new data matrix:

    #output margin from initial boost
    pred <- predict(bst, xgbMat, outputmargin = TRUE)

    #create xgbMatrix with base margin from initial boost
    xgbMatPred <- xgb.DMatrix(data = data,label = label, base_margin = as.matrix(pred))
    watchlist = list(train = xgbMatPred)

Training using the new matrix with the base margin, you can see it picks up where the previous boost finished in terms of log-loss:

    #boost from initial prediction
    bst <- xgb.train(data = xgbMatPred,
                     max_depth = 4, eta = 0.4, nrounds = 50,
                     objective = "multi:softprob", eval_metric = "mlogloss",
                     watchlist = watchlist,
                     num_class = num_class, print_every_n = 10)

    [1]  train-mlogloss:0.015345 
    [11] train-mlogloss:0.014510 
    [21] train-mlogloss:0.013976 
    [31] train-mlogloss:0.013575 
    [41] train-mlogloss:0.013296 
    [50] train-mlogloss:0.013115

However, when I use the same new matrix and parameters in a cross-validation, it seems to ignore the base margin and just starts again from scratch:

    #attempt CV from initial prediction
    set.seed(1)
    bstCV <- xgb.cv(data = xgbMatPred,
                  nfold = 5,              
                  max_depth = 4, eta = 0.4, nrounds = 50,
                  objective = "multi:softprob", eval_metric = "mlogloss",
                  # watchlist = watchlist, 
                  showsd = FALSE,
                  num_class = num_class, print_every_n = 10)

    [1]  train-mlogloss:0.642223|test-mlogloss:0.660639
    [11] train-mlogloss:0.043816|test-mlogloss:0.158496 
    [21] train-mlogloss:0.024552|test-mlogloss:0.182579 
    [31] train-mlogloss:0.020419|test-mlogloss:0.199851 
    [41] train-mlogloss:0.018195|test-mlogloss:0.211206 
    [50] train-mlogloss:0.017033|test-mlogloss:0.217913

The log-loss in round 1 doesn't carry on from the initial boost - rather it resets to no base margin at all.

I've been trying numerous alternatives to this (removing the watchlist parameter, changing between multi:softprob and multi:softmax, different eval_metrics, uninstalling/hard-deleting versions of xgboost).  Nothing seems to work.

I'm not sure why this issue would be happening as [#2006](https://github.com/dmlc/xgboost/issues/2006) has fixed the issue, at least for the binary scenario.  Why it would not work for the multiclass I am not sure of.

Kind regards,

### LINKED ISSUE #5525: "label_upper_bound" and "label_lower_bound" can be set, but not gotten
As of XGBoost 88b64c8162ccfdda611590df09e0c97b9e9a400c:

```
> library(xgboost)
> d = xgb.DMatrix(as.matrix(rnorm(50)))
> setinfo(d, "label_lower_bound", rnorm(50))
[1] TRUE
> getinfo(d, "label_lower_bound")
Error in getinfo.xgb.DMatrix(d, "label_lower_bound") :
  getinfo: name must be one of the following
    'label', 'weight', 'base_margin', 'nrow'
```

### LINKED ISSUE #5549: DMatrix `slice` doesn't handle `label_upper_bound`.
The new `survival:aft` objective added `label_upper_bounds` and `label_lower_bounds` fields, which currently are not supported by the slice method of `DMatrix`, hence `cv` method from R and Python are broken for `survival:aft`.

### LINKED ISSUE #5550: Ranking weight is not handled by `mkgroupfold`.
The Python function `mkgroupfold` tries to handle group info for ranking task, the I couldn't find any logic related to handling per-group weights.  So I'm assuming it's broken.

### PR DESCRIPTION
**Pull request by trivialfis**
Closes #5525, Closes #5549, Closes #5550, Closes #5288 .

### CODE REVIEW

File: src/data/data.cc [hunk 1]
```diff
@@ -205,6 +205,53 @@ void MetaInfo::LoadBinary(dmlc::Stream *fi) {
...
+std::vector<T> Gather(const std::vector<T> &in, common::Span<int const> ridxs, size_t stride = 1) {
+  if (in.empty()) {
+    return {};
+  }
+  auto size = ridxs.size();
+  std::vector<T> out(size * stride);
+  for (auto i = 0ull; i < size; i++) {
+    auto ridx = ridxs[i];
+    for (size_t j = 0; j < stride; ++j) {
+      out.at(i * stride +j) = in.at(ridx * stride + j);
```
Comment (trivialfis) [inline src/data/data.cc:218]: Don't use `at`.

File: R-package/tests/testthat/test_dmatrix.R [hunk 2]
```diff
@@ -50,6 +50,14 @@ test_that("xgb.DMatrix: getinfo & setinfo", {
   labels <- getinfo(dtest, 'label')
   expect_equal(test_label, getinfo(dtest, 'label'))
 
+  expect_true(setinfo(dtest, 'label_lower_bound', test_label))
+  labels <- getinfo(dtest, 'label')
```
Comment (hcho3) [inline R-package/tests/testthat/test_dmatrix.R:54]: I don't think we need this line. `labels` is not used in the test.

**Comment by RAMitchell** [comment]
There are some pros and cons of doing it this way vs the adapter. The adapter way ensured that all dmatrix construction took the same path, so if you added or modified something that needed to be constructed you would only have to deal with it in one place. The con is that it assumes that SparseBatch is available currently and im not sure how to extend it to DeviceQuantileDMatrix. The adapter is also a bit clunky for accessing metainfo.

**Comment by trivialfis** [comment]
@hcho3 Yup.  Will push the fix in a minute.  I just tried extending the adapter and failed.  Because `base_margin_` doesn't equal to the size of data batch, instead it equals to `ndata * mclasses`, so the universal interface created by @RAMitchell is broken with slice.
